transcom/mymove
# MB-1976 Limit concurrent sessions per app

One of the STIG requirements is to limit the number of active sessions
a user can have at the same time. To keep things simple, we are
setting the limit to one active session. However, since there are
three different apps (mil, admin, office), we want to allow people
to be signed in at the same time in each app, but only one active
session per app.

The simplest way to achieve this, while also making it easy to revoke
individual sessions via the admin interface (implemented in a separate
PR), was to add 3 new fields to the Users table, one for each app's
current session ID.

Two other solutions that were considered were: 1) to add a
`current_session_id` field to each of the users, admin_users, and
office_users tables, and 2) creating a new sessions table with a
`user_id`, `session_id`, and `app_name` fields, but both of those
would have added unnecessary complexity.

**Setup**
`make redis_pull`
`make dev_db_migrate`

**To test this feature:**

1. Sign in to http://milmovelocal:3000 in one browser
2. Make note of your `mil_session_token` value by inspecting cookies
in the developer tools
3. Verify that this same value exists in the Users table under the
field `current_mil_session_id`
4. Run `redis-dev`, then `KEYS *`, and verify that this same value
exists
5. Sign in to http://milmovelocal:3000 in another browser
6. Verify that you see "concurrent session detected. Will delete
previous session." in the logs
7. Refresh the first browser
8. Verify that you are no longer logged in
9. Verify that the session ID is deleted from both the DB and Redis
10. Stop the server
11. Set `export FEATURE_FLAG_ROLE_BASED_AUTH=true` in your
`.envrc.local`
12. Run `direnv allow`, then `make server_run`
13. Repeat steps 1-9
14. Sign into milmovelocal:3000, adminlocal:3000 and officelocal:3000
at the same time.
15. Refresh each browser tab.
16. Verify that you can sign into each app once at the same time

## Description

Explain a little about the changes at a high level.

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
echo "Code goes here"
```

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/backend#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:

## References

* [Jira story](tbd) for this change
* [this article](tbd) explains more about the approach used.

## Screenshots

If this PR makes visible UI changes, an image of the finished UI can help reviewers and casual
observers understand the context of the changes. A before image is optional and
can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow instead of using multiple images. You may want to use GIPHY CAPTURE for this! 📸

_Please frame screenshots to show enough useful context but also highlight the affected regions._
